chain-abci/src/app/app_init.rs

```diff
@@ -70,51 +72,117 @@ pub struct ChainNodeApp {
-        last_app_state: &ChainNodeState,
-    ) -> (
-        BTreeMap<StakedStateAddress, TendermintVotePower>,
-        BTreeMap<StakedStateAddress, PubKey>,
-    ) {
-        let mut validator_voting_power = BTreeMap::new();
-        let mut validator_pubkeys = BTreeMap::new();
-        for node in last_app_state.council_nodes.iter() {
-            let pk = ChainNodeApp::get_validator_key(&node);
-            validator_pubkeys.insert(node.staking_account_address, pk);
-            let account = get_account(
-                &node.staking_account_address,
-                &last_app_state.last_account_root_hash,
-                accounts,
-            )
-            .expect("council node staking account should be in the account state");
-            if account.bonded < last_app_state.required_council_node_stake {
-                validator_voting_power.insert(
-                    node.staking_account_address,
-                    TendermintVotePower::from(Coin::zero()),
-                );
-            } else {
-                validator_voting_power.insert(
-                    node.staking_account_address,
-                    TendermintVotePower::from(account.bonded),
-                );
-            }
+fn check_and_store_consensus_params(
+    init_consensus_params: Option<&ConsensusParams>,
+    _validators: &[CouncilNode],
+    _network_params: &InitNetworkParameters,
+    inittx: &mut DBTransaction,
+) {
+    match init_consensus_params {
+        Some(cp) => {
+            // TODO: check validators only used allowed key types
+            // TODO: check unbonding period == cp.evidence.max_age
+            // NOTE: cp.evidence.max_age is currently in the number of blocks
+            // but it should be migrated to "time", in which case this check will make sense
+            // (as unbonding time is in seconds, not blocks)
+            warn!("consensus parameters not checked (TODO)");
+            inittx.put(
+                COL_EXTRA,
+                b"init_chain_consensus_params",
+                &(cp as &dyn Message)
+                    .write_to_bytes()
+                    .expect("consensus params"),
+            );
+        }
+        None => {
+            info!("consensus params not in the initchain request");
         }
-        (validator_voting_power, validator_pubkeys)
     }
+}
+
+fn get_voting_power(
+    distribution: &BTreeMap<RedeemAddress, (Coin, AccountType)>,
+    node_address: &StakedStateAddress,
+) -> TendermintVotePower {
+    match node_address {
+        StakedStateAddress::BasicRedeem(a) => TendermintVotePower::from(distribution[a].0),
+    }
+}
 
+fn store_valid_genesis_state(
+    genesis_time: Timespec,
+    genesis_app_hash: H256,
+    rewards_pool: RewardsPoolState,
+    network_params: InitNetworkParameters,
+    last_account_root_hash: StarlingFixedKey,
+    council_nodes: Vec<CouncilNode>,
+    inittx: &mut DBTransaction,
+) -> ChainNodeState {
+    let last_state = ChainNodeState {
```

Comment:
maybe we can separate as "new" constructor.

Reply:
yes, that's some old code -- `ChainNodeState` may change a bit, so it'll make sense;
I'll add it

Reply:
ok